components/sha2017/sha2017_ota.c

```diff
@@ -72,417 +63,503 @@ static const char *REQUEST = "GET " BADGE_OTA_WEB_PATH " HTTP/1.0\r\n"
-  if ((ret = mbedtls_ssl_set_hostname(&ssl, BADGE_OTA_WEB_SERVER)) != 0) {
-    ESP_LOGE(TAG, "mbedtls_ssl_set_hostname returned -0x%x", -ret);
-    abort();
-  }
+int
+mbedtls_ssl_handshake_(mbedtls_ssl_context *ssl)
+{
+	while (1)
+	{
+		int ret = mbedtls_ssl_handshake(ssl);
+
+		if (ret == MBEDTLS_ERR_SSL_WANT_READ) {
+			// FIXME: implement wait?
+			ESP_LOGI(TAG, "mbedtls_ssl_handshake returned MBEDTLS_ERR_SSL_WANT_READ");
+
+		} else if (ret == MBEDTLS_ERR_SSL_WANT_WRITE) {
+			// FIXME: implement wait?
+			ESP_LOGI(TAG, "mbedtls_ssl_handshake returned MBEDTLS_ERR_SSL_WANT_WRITE");
+
+		} else {
+			return ret;
+		}
+	}
+}
 
-  ESP_LOGI(TAG, "Setting up the SSL/TLS structure...");
+int
+mbedtls_ssl_read_(mbedtls_ssl_context *ssl, unsigned char *buf, size_t len)
+{
+	while (1)
+	{
+		int ret = mbedtls_ssl_read(ssl, buf, len);
+
+		if (ret == MBEDTLS_ERR_SSL_WANT_READ) {
+			// FIXME: implement wait?
+			ESP_LOGI(TAG, "mbedtls_ssl_read returned MBEDTLS_ERR_SSL_WANT_READ");
+
+		} else if (ret == MBEDTLS_ERR_SSL_WANT_WRITE) {
+			// FIXME: implement wait?
+			ESP_LOGI(TAG, "mbedtls_ssl_read returned MBEDTLS_ERR_SSL_WANT_WRITE");
+
+		} else {
+			return ret;
+		}
+	}
+}
 
-  if ((ret = mbedtls_ssl_config_defaults(&conf, MBEDTLS_SSL_IS_CLIENT,
-                                         MBEDTLS_SSL_TRANSPORT_STREAM,
-                                         MBEDTLS_SSL_PRESET_DEFAULT)) != 0) {
-    ESP_LOGE(TAG, "mbedtls_ssl_config_defaults returned %d", ret);
-    task_fatal_error();
-  }
+int
+mbedtls_ssl_write_(mbedtls_ssl_context *ssl, const unsigned char *buf, size_t len)
+{
+	int total_len = 0;
+	while (len > 0)
+	{
+		int ret = mbedtls_ssl_write(ssl, buf, len);
+		if (ret > 0) {
+			total_len += ret;
+			buf = &buf[ret];
+			len -= ret;
+
+		} else if (ret == MBEDTLS_ERR_SSL_WANT_READ) {
+			// FIXME: implement wait?
+			ESP_LOGI(TAG, "mbedtls_ssl_write returned MBEDTLS_ERR_SSL_WANT_READ");
+
+		} else if (ret == MBEDTLS_ERR_SSL_WANT_WRITE) {
+			// FIXME: implement wait?
+			ESP_LOGI(TAG, "mbedtls_ssl_write returned MBEDTLS_ERR_SSL_WANT_WRITE");
+
+		} else {
+			return ret;
+		}
+	}
+
+	return total_len;
+}
 
-  mbedtls_ssl_conf_authmode(&conf, MBEDTLS_SSL_VERIFY_REQUIRED);
-  mbedtls_ssl_conf_ca_chain(&conf, &cacert, NULL);
-  mbedtls_ssl_conf_rng(&conf, mbedtls_ctr_drbg_random, &ctr_drbg);
+static void
+sha2017_ota_task(void *pvParameter)
+{
+	esp_err_t err;
+
+	ESP_LOGI(TAG, "Starting OTA update ...");
+
+	/* determine partitions */
+	const esp_partition_t *part_running = esp_ota_get_running_partition();
+	ESP_LOGI(TAG, "Running from partition type %d subtype %d (offset 0x%08x)",
+			part_running->type, part_running->subtype, part_running->address);
+
+	const esp_partition_t *part_update = esp_ota_get_next_update_partition(NULL);
+	ESP_LOGI(TAG, "Writing to partition type %d subtype %d (offset 0x%08x)",
+			part_update->type, part_update->subtype, part_update->address);
+	assert(part_update != NULL);
```

Comment:
codacy is right: doing `assert(part_update != NULL);` *after* `part_update->type` on the previous line doesn't make sense :)